src/Components/WebView/WebView/src/WebViewManager.cs

```diff
@@ -112,6 +112,23 @@ public Task AddRootComponentAsync(Type componentType, string selector, Parameter
         }
     }
 
+    /// <summary>
+    /// Calls the specified <paramref name="action"/> and passes in the scoped services available to Blazor components.
+    /// </summary>
+    /// <param name="action">The action to invoke.</param>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="action"/> is <c>null</c>.</exception>
+    /// <exception cref="InvalidOperationException">Thrown if this method is called before Blazor has started in the web view.</exception>
+    public void Dispatch(Action<IServiceProvider> action)
+    {
+        ArgumentNullException.ThrowIfNull(action);
+        if (_currentPageContext is null)
+        {
+            throw new InvalidOperationException($"{nameof(Dispatch)} can't be called until after Blazor has started in the web view.");
```

Comment:
What's the likelihood that people run into this through no fault of their own? In general I'm not sure there's an easy way to know if a given BlazorWebView has loaded a PageContext yet. How often will people be calling this as part of some initialization flow and risk race condition bugs?

Ways we could address this:

 * Rename it to `TryDispatch` and just no-op if there's no page context (returning a bool to indicate what happened)
 * Change it to `DispatchAsync(action, cancellationToken)` so it can wait until a page context attaches, albeit with the risk that maybe it never happens

If we think it's likely people will use this as part of initialization logic then the second option is probably more useful. However I appreciate the semantics of this method are now starting to seem a bit complicated.

Reply:
Alright so trying to combine the thoughts from both comment threads into one seems a bit problematic. If we do the `TryDispatch` pattern then we need to return a `bool` but if we also support `Func<T>` then there isn't a great place to return the `bool` (because it would be returning `T`).

Doing the `DispatchAsync` that queues callbacks is quite a bit of code for a value that isn't particularly clear.

These APIs we can add very easily:

```c#
    // (A) run user code, throw if not callable
    (A.1) public void Dispatch(Action<IServiceProvider> action);
    (A.2) public Task DispatchAsync(Action<IServiceProvider> action);
    (A.3) public Task<T> DispatchAsync(Func<IServiceProvider, T> action);

    // (B) run user code, return false if not callable
    (B.1) public bool TryDispatch(Action<IServiceProvider> action);
    (B.2) public Task<bool> TryDispatchAsync(Action<IServiceProvider> action);
```

But anything that requires queueing is non-trivial (a lot of ugly code), or if we want to do sync-over-async it's impossible-ish, or returning success `bool` + return type (possible, but ugly).

So I propose that for now we do all the async (A) and (B) methods (that is, A.2, A.3, B.2) because they are straight-forward to implement and provide a lot of value. I don't have a great answer for the case where you want to _know_ that your call will eventually succeed, but perhaps if in the future we see that is a concern then we can address it at that time, depending on the actual scenario?

Reply:
Avoiding the queuing definitely sounds reasonable. If some clear use case emerges later we can always add it.

What about if we just did B2?

 * Is there really any value in having the A2 variant that's exactly equivalent except it might throw? People can always throw on their own if they get back `false`.
 * As for A3 I get the awkwardness of returning both a bool and a possible return value. We really want something like `Task<Option<T>>`, which isn't idiomatic for .NET. We do have the option to simply not support return values, so people who want to get back a result are responsible for writing it to some closure-captured local. Sorry that conflicts with my earlier suggestion.

I'm not totally committed to either outcome here, but would suggest that if this area is tricky for us to reason about, app developers are definitely not going to anticipate the cases where the call may throw, so it could be safer to limit ourselves to B2, then it's obvious from the naming we're not promising your callback will run.

Reply:
Alright in order to get this into customer hands let's stick with "Less is more" and do only B.2: `public Task<bool> TryDispatchAsync(Action<IServiceProvider> action);` We have a fair bit of time in the .NET 8 time frame to improvise, adapt, and overcome.